### PR TITLE
Add a query cache for dense local DefIds

### PR DESCRIPTION
**Pull request by Zoxc**
Waiting on https://github.com/rust-lang/rust/pull/68988 to land for a perf run.

r? @eddyb

### CODE REVIEW

**Comment by Centril** [comment]
Looks like perf was regressed?

File: src/librustc/ty/query/caches.rs [hunk 1]
```diff
@@ -54,13 +58,18 @@ pub(crate) trait QueryCache<K, V>: Default {
 pub struct DefaultCacheSelector;
 
 impl<K: Eq + Hash, V: Clone> CacheSelector<K, V> for DefaultCacheSelector {
-    type Cache = DefaultCache;
+    type Cache = DefaultCache<()>;
 }
 
-#[derive(Default)]
-pub struct DefaultCache;
+pub struct DefaultCache<D>(PhantomData<D>);
```
Comment (cjgillot) [inline src/librustc/ty/query/caches.rs]: What does this generic parameter `D` do?

Comment (Zoxc) [reply src/librustc/ty/query/caches.rs]: Avoids compiler errors for the `pub type LocalDenseDefIdCacheSelector<V>` alias.

**Comment by Dylan-DPC** [comment]
@zoxc can you resolve conflicts? we can get this reviewed

**Comment by eddyb** [comment]
The problem with this PR was the perf regressions, not reviews.

I suggested a compact query cache on Discord, not sure if @Zoxc has tried that yet.

**Comment by Zoxc** [comment]
@eddyb Is this not what you had in mind?

**Comment by eddyb** [comment]
@Zoxc Oops, I failed to read which PR this was on, for some reason I thought this one landed.

**Review by eddyb — APPROVED** [review]

**Comment by eddyb** [comment]
r=me after rebase

**Comment by Zoxc** [comment]
Perf is still an issue with this one though =P

File: src/librustc/query/mod.rs [hunk 2]
```diff
@@ -75,6 +75,7 @@ rustc_queries! {
         // This can be conveniently accessed by methods on `tcx.hir()`.
         // Avoid calling this query directly.
         query hir_owner(key: DefId) -> &'tcx HirOwner<'tcx> {
```
Comment (eddyb) [inline src/librustc/query/mod.rs:77]: Change the query keys to `LocalDefId`, then you won't have any `is_local` checks, maybe that helps?

Comment (Zoxc) [reply src/librustc/query/mod.rs:77]: They are `LocalDefId` in `master` now. I don't expect removing a single branch to help much though.

**Comment by nikomatsakis** [comment]
Closing this pull request as Zoxc is stepping back from compiler development; see rust-lang/team#316.

**Pull request closed without merging**